### DISCUSSION TITLE
Add support for Gemini & Mistral LLM's

### DISCUSSION THREAD

**Discussion by acherla** [StreetLamb/tribe#123] [Ideas]
Add support for the following LLM's leveraging langchains chat model API for:
- Gemini - https://python.langchain.com/v0.2/docs/integrations/chat/google_generative_ai/
- Mistral - https://python.langchain.com/v0.2/docs/integrations/chat/mistralai/

Both should support:
1. Defining the baseurl/endpoint url path for proxied URL's
2. Defining client/header options
3. Defining the LLM sub model to use (IE gemini-pro, gemini-1.5-flash etc...)

Ill take up this PR

**Comment by acherla** [comment]
I currently have this working in my local build, can create a PR to merge in the implementation.  I will likely add others depending on popularity/requests.

**Reply by StreetLamb** [reply]
Thanks, good for users to have more options. Previously, I had issues with Gemini due to it not supporting System Messages. I will gladly review and merge it if you raise a PR.

**Reply by acherla** [reply]
Seems like gemini now supports System messages but only after upgrading to the latest client 1.0.10, which requires langcore 2.33

https://python.langchain.com/v0.2/api_reference/google_genai/chat_models/langchain_google_genai.chat_models.ChatGoogleGenerativeAI.html#langchain_google_genai.chat_models.ChatGoogleGenerativeAI.client_options 

```
template = ChatPromptTemplate.from_messages(
    [("system", "You are Cat Agent 007"), ("human", "{question}")]
).with_config({"run_name": "my_template", "tags": ["my_template"]})
```

**Reply by StreetLamb** [reply]
Just tried it out locally and seems it's having some issues with tool calling: 
```
langchain_google_genai.chat_models.ChatGoogleGenerativeAIError: Invalid argument provided to Gemini: 400 * GenerateContentRequest.tools[0].function_declarations[0].name: Invalid function name. Must start with a letter or an underscore. Must be alphameric (a-z, A-Z, 0-9), underscores (_), dots (.) or dashes (-), with a maximum length of 64.
```

Looks like others are also experiencing it: https://github.com/langchain-ai/langchain/issues/24225

**Reply by acherla** [reply]
Yea i had to update your get_tool_definition in members.py to get it to work as gemini uses a slightly different api spec for making function calls.  If you want I can move up the PR for this as I have a working implementation.  

See updates below for get_tool_definition that I applied: added function.name and function.parameters.task.types.

```
    def get_tool_definition(self, options: list[str]) -> dict[str, Any]:
        """Return the tool definition to choose next team member and provide the task."""
        return {
            "type": "function",
            "name": "route",
            "function": {
                "name": "route",
                "description": (
                    "Provide both a task and the next most appropriate team member to perform it."
                    "\n'next' - The team member you should call."
                    "\n'task' - The task given to the team member."
                    "\nYou must provide both 'task' and 'next'."
                    "\n\nExample:"
                    "\nQn: How to cook food?"
                    '\n{"task": "Provide cooking instructions", "next": "CookingExpert"}'
                    "\n\nQn: How do you play soccer?"
                    '\n{"task": "Provide advice to play soccer", "next": "SoccerTeam"}'
                    "\n\nQn: How to make a dog happy?"
                    "\nAns: Pat its head and rub its belly"
                    '\n{"task": "No further tasks", "next": "FINISH"}'
                ),
                "parameters": {
                    "title": "routeSchema",
                    "type": "object",
                    "properties": {
                        "task": {
                            "title": "task",
                            "description": "Provide the next task only if answer is still incomplete. Else say no further task.",
                            "type": "object",
                        },
                        "next": {
                            "title": "next",
                            "type": "object",
                            "description": "Choose the next most appropriate team member if answer is still incomplete. Else choose FINISH.",
                            "anyOf": [
                                {"enum": options},
                            ],
                        },
                    },
                    "required": ["next", "task"],
                },
            },
        }
```

I am thinking it might make sense to make an abstract class that is extendable to different model types to handle changes in function tool request api calls by model

**Reply by StreetLamb** [reply]
Hmm I'm still having issues on my side when using it in hierarchical workflows. No errors but the leader is not routing to other members. Perhaps after you raise the PR I can try agin.
